### ISSUE TITLE
Onnx importer workarounds

### ISSUE THREAD

**Issue by cscribano** [kendryte/nncase#117]
Congrats for the amazing work!

I want to higlight some potential bugfixes for the onnx parser.

My configuration:
 * Pytorch: 1.2.0 (Onnx Opset 10)
 * Input shape: (1, 3, 28, 28) (for all the following examples)

All the reference ONNX models are available: [onnx_examples.zip](https://github.com/kendryte/nncase/files/4517249/onnx_examples.zip)

### AvgPool2d
(This can be probably applied to any Pool2d)\
Minimal example:
```
class Network_Pool(torch.nn.Module):
       def __init__(self):
           super(Network_Pool, self).__init__()
           self.a = torch.nn.AvgPool2d(7)
   
       def forward(self, x):
           x = self.a(x)
           return x
```
(example file: pool.onnx)

After exporting to onnx I get a segfault when triyng to parse the model in nncase,
the problem appears to originate from in _ops/pool.cpp:80_. The function fails for some reason 
to obtain the attribute **kernel_shape** (return -1).
My workaround so far was to use **strides** for kernel_shape (I'v never seen used pooling operations with stride != Kernel size)
this is surely not the solution but can do the trick before figuring out the root cause of the issue.

_ops/pool.cpp:113-144_
```
pads[0] = get_windowed_padding(input_shape[2], strides[0], strides[0], dilations[0], true);
pads[1] = get_windowed_padding(input_shape[3], strides[1], strides[1], dilations[1], true);
```

_ops/pool.cpp:118_
```
auto op { graph_.emplace<reduce_window2d>(reduce_op, move(input_shape), init_value, strides[0], strides[1], pads[0], pads[1], strides[0], strides[1], dilations[0], dilations[1], value_range<float>::full()) };
```

I've also implemented the (missing) test for the Pool operation and this seems to confirm this workaround working
(more investigations are surely needed).

### Reshape (With ONNX-Sim)
Minimal example:
```
class Network_reshape(torch.nn.Module):
    def __init__(self):
        super(Network_reshape, self).__init__()
        self.a = torch.nn.Conv2d(3, 3, 1)

    def forward(self, x):
        x = self.a(x)
        x = torch.reshape(x, (1,3,784))
        return x
```
(example files: reshape.onnx, reshape_sim.onnx)

When exporting from torch.onnx It works as supposed, however after simpyfing the model using (https://github.com/daquexian/onnx-simplifier/)
becomes impossible to import the model (note: OnnxSim comes handy to deal with the unsupported BatchNorm layers).

Onnxsim remove constant nodes, so the target shape becomes an initializer with 
DataType INT64 (7), this cause assert in _onnx_importer.cpp:565_ to fail.
Removing the assert will Not fix the problem: strangely enough _tensor.int64_data_size()_ result always zero
even if the TensorProto object appears to contain the right RAW bytes.

I changed _onnx_importer.cpp:573-589_ as follow:
```
    case TensorProto_DataType_INT64:
    {
        axis_t result;

        //assert(tensor.int64_data_size() > 0);
        if(tensor.int64_data_size() == 0) {
            const size_t size{tensor.raw_data().size() / 8};
            const auto *ptr{reinterpret_cast<const int64_t *>(tensor.raw_data().data())};
            transform(ptr, ptr + size, back_inserter(result),
                      [](const auto e) { return static_cast<float>(e); });
        }

        else {
            transform(tensor.int64_data().begin(), tensor.int64_data().end(), back_inserter(result),
                      [](const auto a) { return a; });
        }
        return result;
```

This way the Raw data gets parsed to an array and in the end the correct result is obtained.
It will be a good idea to test all the opcodes for operability with Onnx-Sim.

### Flatten
The implementation of the Flatten op, while being very simple to model after Reshape, can be very userful since the PyTorch exporter whill cast all the reshape operation to Flatten nodes whenever is possible.

I succesfully implemented this operation, if this can help I can post my implementation.

### Gemm

This operation would be a great addition being used for Fully-Connected layers.
I'll try to implement this by myself when possible.

**Comment by sunnycase** [comment]
Would you mind submitting a pull request?

**Comment by cscribano** [comment]
I need to check against #103

**Comment by cscribano** [comment]
I can confirm that #103 fixes the issues thati I've mentioned with Pool and Reshape, so my proposed workarounds are no more necessary.
I will eventually submit a pr to add the Flatten operation

**Comment by cscribano** [comment]
#139 Added 'Flatten' operation, all the problems mentioned are now solved.
Good work!
